### PR TITLE
Add support for "List build targets" command

### PR DESCRIPTION
**Pull request by ayoub-benali**
Fix https://github.com/scalameta/metals-sublime/issues/78

Peek 2022-09-16 21-03

### CODE REVIEW

File: LSP-metals.sublime-commands [hunk 1]
```diff
@@ -31,5 +31,6 @@
             "base_file": "${packages}/LSP-metals/LSP-metals.sublime-settings",
             "default": "{\n\t$0\n}\n"
         }
-    }
+    },
+    { "caption": "LSP-metals: Show Build Target Info", "command": "lsp_metals_show_build_target_info"}
```
Comment (ayoub-benali) [inline LSP-metals.sublime-commands:35]: This to make the command available in the sublime palette

File: plugin.py [hunk 2]
```diff
@@ -16,17 +16,18 @@
...
     from . commands.lsp_metals_execute_command import LspMetalsExecuteCommand
     from . commands.lsp_metals_file_decoder import LspMetalsFileDecoderCommand
     from . commands.lsp_metals_find_in_dependency import LspMetalsFindInDependencyCommand
     from . commands.lsp_metals_focus import LspMetalsFocusViewCommand, ActiveViewListener
     from . commands.lsp_metals_goto import LspMetalsGoto
-    from . commands.lsp_metals_metals_goto_location import LspMetalsMetalsGotoLocationCommand
     from . commands.lsp_metals_goto_super_method import LspMetalsSendPositionCommand
+    from . commands.lsp_metals_metals_goto_location import LspMetalsMetalsGotoLocationCommand
     from . commands.lsp_metals_new_scala_file import LspMetalsNewScalaFileCommand
+    from . commands.lsp_metals_show_build_target_info import LspMetalsShowBuildTargetInfoCommand
```
Comment (ayoub-benali) [inline plugin.py:28]: The command needs to be explicitly imported

File: commands/lsp_metals_show_build_target_info.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
+from . lsp_metals_text_command import LspMetalsTextCommand
+from . utils import handle_error
+from LSP.plugin.core.protocol import Error
+from LSP.plugin.core.types import List, Any
+
+import sublime
+
+
+class LspMetalsShowBuildTargetInfoCommand(LspMetalsTextCommand):
```
Comment (ayoub-benali) [inline commands/lsp_metals_show_build_target_info.py:9]: The actual command that first list the build targets then invokes the decoding command

File: commands/lsp_metals_file_decoder.py [hunk 4]
```diff
@@ -22,17 +22,22 @@ class LspMetalsFileDecoderCommand(LspMetalsTextCommand):
...
             extension = os.path.splitext(self.view.file_name())[1][1:]
             accepted_extentions = self._decoders.get(decoding_type)
-            return accepted_extentions is not None and extension in accepted_extentions
+            return decoding_type == self._build_target or (accepted_extentions is not None and extension in accepted_extentions)
         else:
             return False
 
-    def run(self, edit: sublime.Edit, decoding_type: str) -> None:
-        uri = 'metalsDecode:' + filename_to_uri(self.view.file_name()) + '.' + decoding_type
+    def run(self, edit: sublime.Edit, decoding_type: str, file_path: str = '') -> None:
```
Comment (ayoub-benali) [inline commands/lsp_metals_file_decoder.py:35]: adapted the existing decoding command to accept the file path as argument

File: commands/lsp_metals_file_decoder.py [hunk 5]
```diff
@@ -22,17 +22,22 @@ class LspMetalsFileDecoderCommand(LspMetalsTextCommand):
...
         }
+    _build_target = 'metals-buildtarget'
 
-    def is_enabled(self, decoding_type: str) -> bool:
+    def is_enabled(self, decoding_type: str, file_path: str = '') -> bool:
         if super().is_enabled(None, None):
             extension = os.path.splitext(self.view.file_name())[1][1:]
             accepted_extentions = self._decoders.get(decoding_type)
-            return accepted_extentions is not None and extension in accepted_extentions
+            return decoding_type == self._build_target or (accepted_extentions is not None and extension in accepted_extentions)
```
Comment (ayoub-benali) [inline commands/lsp_metals_file_decoder.py:31]: And this is to accept the new decoding type

File: commands/lsp_metals_show_build_target_info.py [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+        elif isinstance(response, list) and response:
+            self.view.window().show_quick_panel(
+                response,
+                lambda i: self._on_select(response, i),
+                placeholder="Select the build target to display")
+
+    def _on_select(self, items: List[str], index: int) -> None:
+        if index >= 0:
+            session = self.session_by_name(self.session_name)
+            root = session._workspace_folders[0]
```
Comment (ayoub-benali) [inline commands/lsp_metals_show_build_target_info.py:32]: @rchl @rwols is this ok or is there a better way ?

File: commands/lsp_metals_show_build_target_info.py [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+            session.execute_command(params, progress=True).then(self._handle_response)
+
+    def _handle_response(self, response: Any) -> None:
+        if isinstance(response, Error):
+            handle_error(self._command_name, response)
+        elif isinstance(response, list) and response:
+            self.view.window().show_quick_panel(
+                response,
+                lambda i: self._on_select(response, i),
+                placeholder="Select the build target to display")
```
Comment (rchl) [inline commands/lsp_metals_show_build_target_info.py:27]: You can override `handle_success_async` instead and then you don't need to define duplicated `run`. See implementation of `LspExecuteCommand`: https://github.com/sublimelsp/LSP/blob/566a49e2e4c8f9cfd057861681b0117bbe274bd2/plugin/execute_command.py#L49-L69

File: commands/lsp_metals_show_build_target_info.py (hunk 6 above)
Comment (rchl) [reply commands/lsp_metals_show_build_target_info.py:32]: `get_workspace_folders` is the public API.

File: commands/lsp_metals_show_build_target_info.py (hunk 7 above)
Comment (ayoub-benali) [reply commands/lsp_metals_show_build_target_info.py]: Thanks, I think I can refactor other commands this way as well.

**Review by rchl — APPROVED** [review]

**Pull request merged**